common/clickteam.cpp

```diff
@@ -0,0 +1,306 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "common/array.h"
+#include "common/clickteam.h"
+#include "common/gzio.h"
+#include "common/debug.h"
+#include "common/ptr.h"
+#include "common/substream.h"
+#include "common/memstream.h"
+
+#define STUB_SIZE 0x16000
+#define FLAG_COMPRESSED 1
+
+namespace Common {
+
+ClickteamInstaller::ClickteamFileDescriptor::ClickteamFileDescriptor(const ClickteamTag& contentsTag, uint32 off)
+  : _fileDataOffset(0), _fileDescriptorOffset(0), _compressedSize(0), _uncompressedSize(0) {
+	uint32 stringsOffset = 36;
+	unsigned char *tag = contentsTag._contents + off;
+	uint32 lmax = contentsTag._size - off;
+	if (lmax < 0x24)
+		return;
+	uint16 ls = READ_LE_UINT32(tag), l;
+	if (ls < 0x24)
+		return;
+
+	l = MIN((uint32)ls, lmax);
+
+	uint16 flags = READ_LE_UINT32(tag+22);
+	if (flags & 6)
+		stringsOffset += 0x12;
+	if (flags & 8)
+		stringsOffset += 0x18;
+	if (stringsOffset >= l) {
+		return;
+	}
+	_fileDataOffset = READ_LE_UINT32(tag + 6);
+	_compressedSize = READ_LE_UINT32(tag + 10);
+	_uncompressedSize = READ_LE_UINT32(tag + 14);
+	_expectedCRC = READ_LE_UINT32(tag + 18);
+	char *strings = (char *)tag + stringsOffset;
+	char *p;
+	for (p = strings; p < (char*)tag + lmax && *p; p++);
+	_fileName = Common::String(strings, p - strings);
+	_fileDescriptorOffset = off;
+}
+
+ClickteamInstaller::ClickteamTag* ClickteamInstaller::getTag(ClickteamTagId tagId) const {
+	return _tags.getValOrDefault((uint16) tagId).get();
+}
+
+static uint32 computeCRC(byte* buf, uint32 sz, unsigned int previous)
+{
```

Comment:
the code formatting is off in many places here.  Please, check how the curly braces stick in other instances too.

Reply:
Done